transport/socks5/packet_dialer.go

```diff
@@ -0,0 +1,218 @@
+var udpPool = slicepool.MakePool(clientUDPBufferSize)
+
+type packetConn struct {
+	dstAddr net.Addr
+	pc      net.Conn
+	sc      transport.StreamConn
+}
+
+var _ net.Conn = (*packetConn)(nil)
+
+func (c *packetConn) LocalAddr() net.Addr {
+	return c.pc.LocalAddr()
+}
+
+func (c *packetConn) RemoteAddr() net.Addr {
+	return c.dstAddr
+}
+
+func (c *packetConn) SetDeadline(t time.Time) error {
+	return c.pc.SetDeadline(t)
+}
+
+func (c *packetConn) SetReadDeadline(t time.Time) error {
+	return c.pc.SetReadDeadline(t)
+}
+
+func (c *packetConn) SetWriteDeadline(t time.Time) error {
+	return c.pc.SetWriteDeadline(t)
+}
+
+func (c *packetConn) Read(b []byte) (int, error) {
+	lazySlice := udpPool.LazySlice()
+	buffer := lazySlice.Acquire()
+	defer lazySlice.Release()
+	n, err := c.pc.Read(buffer)
+	if err != nil {
+		return 0, err
+	}
+	// Minimum size of header is 10 bytes
+	// 2 bytes for reserved, 1 byte for fragment, 1 byte for address type, 4 byte for ipv4, 2 bytes for port
+	if n < 10 {
+		return 0, fmt.Errorf("invalid SOCKS5 UDP packet: too short")
+	}
+
+	pkt := buffer[:n]
+
+	// Start parsing the header
+	rsv := pkt[:2]
+	if rsv[0] != 0x00 || rsv[1] != 0x00 {
+		return 0, fmt.Errorf("invalid reserved bytes: expected 0x0000, got %#x%#x", rsv[0], rsv[1])
+	}
+
+	frag := pkt[2]
+	if frag != 0 {
+		return 0, errors.New("fragmentation is not supported")
+	}
+
+	// Do something with address?
+	_, addrLen, err := readAddress(bytes.NewReader(pkt[3:]))
+	if err != nil {
+		return 0, fmt.Errorf("failed to read address: %w", err)
+	}
+
+	// Calculate the start position of the actual data
+	headerLength := 4 + addrLen + 2 // RSV (2) + FRAG (1) + ATYP (1) + ADDR (variable) + PORT (2)
+	if n < headerLength {
+		return 0, fmt.Errorf("invalid SOCKS5 UDP packet: header too short")
+	}
+
+	// Copy the payload into the provided buffer
+	payloadLength := n - headerLength
+	if payloadLength > len(b) {
+		return 0, io.ErrShortBuffer
+	}
+	copy(b, buffer[headerLength:n])
+
+	return payloadLength, nil
+}
+
+func (c *packetConn) Write(b []byte) (int, error) {
+	// Encapsulate the payload in a SOCKS5 UDP packet as specified in
+	// https://datatracker.ietf.org/doc/html/rfc1928#section-7
+	// The minimum preallocated header size (10 bytes)
+	header := make([]byte, 10)
+	header = append(header[:0],
+		0x00, 0x00, // Reserved
+		0x00, // Fragment number
+		// To be appended below:
+		// ATYP, IPv4, IPv6, Domain Name, Port
+	)
+	header, err := appendSOCKS5Address(header, c.dstAddr.String())
+	if err != nil {
+		return 0, fmt.Errorf("failed to append SOCKS5 address: %w", err)
+	}
+	// Combine the header and the payload
+	fullPacket := append(header, b...)
+	return c.pc.Write(fullPacket)
+}
+
+func (c *packetConn) Close() error {
+	return errors.Join(c.sc.Close(), c.pc.Close())
+}
+
+// DialPacket creates a packet [net.Conn] via SOCKS5.
+func (d *Dialer) DialPacket(ctx context.Context, dstAddr string) (net.Conn, error) {
```

Comment:
This should be renamed to `ListenPacket`, not take a destination address and return a `net.PacketConn`.

Reply:
@fortuna I implemented and tested this change.